File: ios-oc/VoiceAgent/Chat/AgentStateView.m
//
//  AgentStateView.m
//  VoiceAgent
//
//  Created by qinhui on 2025/11/17.
//

#import "AgentStateView.h"
#import <Masonry/Masonry.h>

static inline UIColor *VAHexColor(NSUInteger hexValue, CGFloat alpha) {
    return [UIColor colorWithRed:((hexValue >> 16) & 0xFF) / 255.0
                           green:((hexValue >> 8) & 0xFF) / 255.0
                            blue:(hexValue & 0xFF) / 255.0
                           alpha:alpha];
}

@interface AgentStateView ()

@property (nonatomic, strong) UILabel *statusLabel;

@end

@implementation AgentStateView

- (instancetype)initWithFrame:(CGRect)frame {
    self = [super initWithFrame:frame];
    if (self) {
        [self setupUI];
        [self setupConstraints];
        self.hidden = YES;
    }
    return self;
}

- (void)setupUI {
    self.backgroundColor = VAHexColor(0x1E293B, 0.8);
    self.layer.cornerRadius = 12;
    self.layer.borderWidth = 0.5;
    self.layer.borderColor = VAHexColor(0x334155, 0.5).CGColor;
    
    self.statusLabel = [[UILabel alloc] init];
    self.statusLabel.textColor = VAHexColor(0xCBD5E1, 1.0);
    self.statusLabel.font = [UIFont systemFontOfSize:14 weight:UIFontWeightMedium];
    self.statusLabel.textAlignment = NSTextAlignmentCenter;
    [self addSubview:self.statusLabel];
}

- (void)setupConstraints {
    [self.statusLabel mas_makeConstraints:^(MASConstraintMaker *make) {
        make.edges.equalTo(self).inset(8);
    }];
}

- (void)updateState:(NSInteger)state {
    if (state == 5) { // unknown
        self.hidden = YES;
        return;
    }
    
    self.hidden = NO;
    NSString *statusText;
    switch (state) {
        case 0: // idle
            statusText = @"Idle";
            self.backgroundColor = VAHexColor(0x64748B, 0.25);
            break;
        case 1: // silent
            statusText = @"Silent";
            self.backgroundColor = VAHexColor(0x475569, 0.25);
            break;
        case 2: // listening
            statusText = @"Listening";
            self.backgroundColor = VAHexColor(0x10B981, 0.2);
            break;
        case 3: // thinking
            statusText = @"Thinking";
            self.backgroundColor = VAHexColor(0xF59E0B, 0.2);
            break;
        case 4: // speaking
            statusText = @"Speaking";
            self.backgroundColor = VAHexColor(0x3B82F6, 0.2);
            break;
        default:
            statusText = @"";
            self.backgroundColor = VAHexColor(0x1E293B, 0.8);
            break;
    }
    self.statusLabel.text = statusText;
}

@end

File: ios-swift/VoiceAgent/Chat/AgentStateView.swift
//
//  AgentStateView.swift
//  VoiceAgent
//
//  Created by qinhui on 2025/11/17.
//

import UIKit
import SnapKit

class AgentStateView: UIView {
    private let statusLabel = UILabel()

    override init(frame: CGRect) {
        super.init(frame: frame)
        setupUI()
        setupConstraints()
        isHidden = true
    }

    required init?(coder: NSCoder) {
        fatalError("init(coder:) has not been implemented")
    }

    private func setupUI() {
        backgroundColor = AppColors.bgCard
        layer.cornerRadius = 12
        layer.borderWidth = 0.5
        layer.borderColor = AppColors.borderDefault.cgColor

        statusLabel.textColor = AppColors.textSubtitle
        statusLabel.font = .systemFont(ofSize: 14, weight: .medium)
        statusLabel.textAlignment = .center
        addSubview(statusLabel)
    }

    private func setupConstraints() {
        statusLabel.snp.makeConstraints { make in
            make.edges.equalToSuperview().inset(8)
        }
    }

    func updateState(_ state: AgentState) {
        if state == .unknown {
            isHidden = true
            return
        }

        isHidden = false
        let (text, color): (String, UIColor) = {
            switch state {
            case .idle:     return ("Idle", AppColors.stateIdle)
            case .silent:   return ("Silent", AppColors.stateSilent)
            case .listening: return ("Listening", AppColors.stateListening)
            case .thinking: return ("Thinking", AppColors.stateThinking)
            case .speaking: return ("Speaking", AppColors.stateSpeaking)
            case .unknown:  return ("", AppColors.stateIdle)
            @unknown default: return ("", AppColors.stateIdle)
            }
        }()

        statusLabel.text = text
        backgroundColor = color.withAlphaComponent(0.2)
    }
}
